Fix calling wrong controller lifecycle methods

davidalejandroaguilar opened the pull request:
# Bug fix

## Description

When selecting the controller for a specific reflex, we were always selecting the first one. This caused us to run the lifecycle methods of the wrong controller if the user specified other controllers via `data-controller`. This commit finds the correct controller and if not found, it uses the default stimulus controller as before.

Some pseudocode examples of what was happening:

Before:

```js
// Given these controllers (identifiers here)
const controllers = ['sortable', 'ui', 'autogrow', 'test']
// When reflex name was
const reflex = 'click->TestReflex#create'
// We were doing
controllers[0]
// Which selected the wrong controller. i.e. 'sortable' instead of 'test'
```

```js
// Given these controllers (identifiers here)
const controllers = ['sortable', 'ui', 'autogrow', 'test']
// When reflex name was
const reflex = 'click->TestReflex#create'
// We now do
controllers.find(controller => controller.identifier === reflexName)
// Which selects the correct controller. i.e. 'test'
```

Fixes https://github.com/hopsoft/stimulus_reflex/issues/225

## Checklist

- [x] My code follows the style guidelines of this project
- [x] Checks (StandardRB & Prettier-Standard) are passing

## Notes

I'm marking this as WIP as I'm not sure which branch this should go to, or other build steps.

leastbad commented on javascript/utils.js:
```diff
@@ -32,3 +32,9 @@ export const debounce = (callback, delay = 250) => {
 }
+
+export const extractReflexName = reflexString => {
+  const match = reflexString.match(/(.*->)?(.*)Reflex/)
```
Heads-up: we've made the trailing `Reflex` string fragment optional, because it's redundant. See https://github.com/hopsoft/stimulus_reflex/blob/895ee97bcdbe8d80836f177d47a1668be041f61e/lib/stimulus_reflex/channel.rb#L25

leastbad commented:
@davidalejandroaguilar I'd love to get this merged this weekend, but you have to update the regex to make it accept both "click->CommentReflex#post" AND "click->Comment#post".

davidalejandroaguilar commented:
Hey @leastbad , I've modified the regex to account for these cases:

* click->TestReflex#create
* TestReflex#create
* click->Test#create
* Test#create

I added a basic Regex explanation within the commit as well. Let me know if I'm missing something or if you need me to reorganize /squash commits.

Sorry for the huge delay, it's been very busy weeks and I'm struggling to find time to devote to projects I like.

jasoncharnes commented:
We're using custom controllers that don't match the reflex name:

- **Ruby**: `app/reflexes/cms/campaigns_reflex.rb`
- **Stimulus**: `app/javascript/cms/campaigns/edit_controller.js`

When I point our libraries to `master` we lose all the lifecycle methods we defined because it doesn't see our controller now.

Would it be acceptable for me to open a PR that falls back to `controllers[0]` if a match isn't found?

leastbad commented:
@jasoncharnes that **sucks**. Let's get you sorted out.

So I assume that Ruby doesn't really care where the Reflex is, but Stimulus is dealing with a `campaigns--edit` controller? Is there really nothing we can do to be super smart about this?

In the meantime, PR please!
